File: Pascal's Triangle II.cpp
/*
只需要两个vector即可，
用v来进行计算，用v_last来存上一行的
rowIndex是从0th开始算的
*/
class Solution {
public:
     vector<int> getRow( int rowIndex) {
        vector< int> v, v_last;
        rowIndex++; //先加1，从1开始算，这样行数和元素数相等
        v.push_back(1);
        v_last = v;
        if(rowIndex == 1) return v;
        for( int i=2; i<=rowIndex; i++){ //从第2行到第rowIndex行
            v.clear();
            v.push_back(1);
            for( int j=0; j<i-2; j++) { //中间有i-2个元素
                v.push_back(v_last[j]+v_last[j+1]);
            }
            v.push_back(1);
            v_last = v;
        }
        return v;
    }
};

File: java/Pascal's Triangle II.java
public class Solution {
    public ArrayList<Integer> getRow(int rowIndex) {
        ArrayList<Integer> res = new ArrayList<Integer>();
        ArrayList<Integer> res_next = new ArrayList<Integer>();
        res.add(1);
        if(rowIndex == 0) return res;
        res.add(1);
        if(rowIndex == 1) return res;
        for(int i=2; i<=rowIndex; i++) {
            res_next.add(1);
            for(int j=1; j<i; j++) {
                res_next.add(res.get(j-1) + res.get(j));
            }
            res_next.add(1);
            res.clear();
            res.addAll(res_next);
            res_next.clear();
        }
        return res;
    }
}public class Solution {
    public int uniquePaths(int m, int n) {
        int v[][] = new int[m][n];
        for(int i=0; i<m; i++) {
            v[i][0] = 1;
        }
        for(int j=0; j<n; j++) {
            v[0][j] = 1;
        }
        for(int i=1; i<m; i++) {
            for(int j=1; j<n; j++) {
                v[i][j] = v[i-1][j] + v[i][j-1];
            }
        }
        return v[m-1][n-1];
    }
}